Implement clearcoat per the Filament and the `KHR_materials_clearcoat` specifications.

pcwalton opened the pull request:
Clearcoat is a separate material layer that represents a thin translucent layer of a material. Examples include (from the [Filament spec]) car paint, soda cans, and lacquered wood. This commit implements support for clearcoat following the Filament and Khronos specifications, marking the beginnings of support for multiple PBR layers in Bevy.

The [`KHR_materials_clearcoat`] specification describes the clearcoat support in glTF. In Blender, applying a clearcoat to the Principled BSDF node causes the clearcoat settings to be exported via this extension. As of this commit, Bevy parses and reads the extension data when present in glTF. Note that the `gltf` crate has no support for `KHR_materials_clearcoat`; this patch therefore implements the JSON semantics manually.

Clearcoat is integrated with `StandardMaterial`, but the code is behind a series of `#ifdef`s that only activate when clearcoat is present. Additionally, the `pbr_feature_layer_material_textures` Cargo feature must be active in order to enable support for clearcoat factor maps, clearcoat roughness maps, and clearcoat normal maps. This approach mirrors the same pattern used by the existing transmission feature and exists to avoid running out of texture bindings on platforms like WebGL and WebGPU. Note that constant clearcoat factors and roughness values *are* supported in the browser; only the relatively-less-common maps are disabled on those platforms.

This patch refactors the lighting code in `StandardMaterial` significantly in order to better support multiple layers in a natural way. That code was due for a refactor in any case, so this is a nice improvement.

A new demo, `clearcoat`, has been added. It's based on [the corresponding three.js demo], but all the assets (aside from the skybox and environment map) are my original work.

[Filament spec]: https://google.github.io/filament/Filament.html#materialsystem/clearcoatmodel

[`KHR_materials_clearcoat`]: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_clearcoat/README.md

[the corresponding three.js demo]: https://threejs.org/examples/webgl_materials_physical_clearcoat.html

![Screenshot 2024-04-19 101143](https://github.com/bevyengine/bevy/assets/157897/3444bcb5-5c20-490c-b0ad-53759bd47ae2)

![Screenshot 2024-04-19 102054](https://github.com/bevyengine/bevy/assets/157897/6e953944-75b8-49ef-bc71-97b0a53b3a27)

## Changelog

### Added

* `StandardMaterial` now supports a clearcoat layer, which represents a thin translucent layer over an underlying material.
* The glTF loader now supports the `KHR_materials_clearcoat` extension, representing materials with clearcoat layers.

## Migration Guide

* The lighting functions in the `pbr_lighting` WGSL module now have clearcoat parameters, if `STANDARD_MATERIAL_CLEARCOAT` is defined.

* The `R` reflection vector parameter has been removed from some lighting functions, as it was unused.

robtfm commented on crates/bevy_gltf/src/loader.rs:
```diff
@@ -1636,6 +1692,104 @@ struct AnimationContext {
+
+impl ClearcoatExtension {
+    #[allow(unused_variables)]
+    fn parse(
```
could this be upstreamed?

pcwalton replied:
I can, but `gltf` hasn't had any updates for a month. Do you want to block this PR on that?

robtfm commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -170,6 +170,34 @@ fn pbr_input_from_standard_material(
+#ifdef VERTEX_UVS
+#ifdef PBR_MULTI_LAYER_MATERIAL_TEXTURES_SUPPORTED
+        if ((pbr_bindings::material.flags & pbr_types::STANDARD_MATERIAL_FLAGS_CLEARCOAT_TEXTURE_BIT) != 0u) {
+#ifdef MESHLET_MESH_MATERIAL_PASS
+            pbr_input.material.clearcoat *= textureSampleGrad(pbr_bindings::clearcoat_texture, pbr_bindings::clearcoat_sampler, uv, in.ddx_uv, in.ddy_uv).r;
+#else   // MESHLET_MESH_MATERIAL_PASS
+            pbr_input.material.clearcoat *= textureSampleBias(pbr_bindings::clearcoat_texture, pbr_bindings::clearcoat_sampler, uv, view.mip_bias).r;
+#endif  // MESHLET_MESH_MATERIAL_PASS
```
variants of this block are repeated 10 times now, could we factor it out?

pcwalton replied:
Done.

robtfm commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -69,19 +69,9 @@ fn apply_normal_mapping(
     world_normal: vec3<f32>,
     double_sided: bool,
     is_front: bool,
-#ifdef VERTEX_TANGENTS
```
just calling out for other reviewers - this function is no longer called unless VERTEX_UVS / VERTEX_TANGENTS are active so the behaviour changes in the non-tangent case don't matter.

but also @pcwalton - array_texture.wgsl needs updating for this api change.

pcwalton replied:
Good catch, fixed.

robtfm commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -241,7 +241,7 @@ fn apply_pbr_lighting(
             transmitted_shadow = shadows::fetch_point_shadow(light_id, diffuse_transmissive_lobe_world_position, -in.world_normal);
         }
-        let transmitted_light_contrib = lighting::point_light(diffuse_transmissive_lobe_world_position.xyz, light_id, 1.0, 1.0, -in.N, -in.V, vec3<f32>(0.0), vec3<f32>(0.0), vec2<f32>(0.1), diffuse_transmissive_color);
+        let transmitted_light_contrib = lighting::point_light(diffuse_transmissive_lobe_world_position.xyz, light_id, 1.0, 1.0, -in.N, -in.V, vec3<f32>(0.0), vec3<f32>(0.0), vec2<f32>(0.1), 0.0, 0.0, diffuse_transmissive_color);
```
this doesn't look right - you #ifdef'd 5 params in the point_light call above.

robtfm commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -223,7 +205,25 @@ fn apply_pbr_lighting(
             shadow = shadows::fetch_point_shadow(light_id, in.world_position, in.world_normal);
         }
-        let light_contrib = lighting::point_light(in.world_position.xyz, light_id, roughness, NdotV, in.N, in.V, R, F0, f_ab, diffuse_color);
+        let light_contrib = lighting::point_light(
+            in.world_position.xyz,
+            light_id,
+            roughness,
+            NdotV,
+            in.N,
+            in.V,
+            R,
+            F0,
+            f_ab,
+#ifdef STANDARD_MATERIAL_CLEARCOAT
+            clearcoat_NdotV,
+            clearcoat_N,
+            clearcoat_R,
+            clearcoat,
+            clearcoat_roughness,
+#endif  // STANDARD_MATERIAL_CLEARCOAT
+            diffuse_color
```
maybe we could have a struct to hold all these arguments ... or most of them, so we could reuse it with a couple extra params for all the lighting calls?

pcwalton replied:
Working on this. This is a huge change :)

pcwalton replied:
This is now done. The entire `pbr_lighting.wgsl` file has been refactored from top to bottom to use structures throughout. I also changed `environment_map.wgsl` to do the same

robtfm commented on crates/bevy_pbr/src/pbr_material.rs:
```diff
@@ -753,6 +821,20 @@ impl AsBindGroupShaderType<StandardMaterialUniform> for StandardMaterial {
             }
         }
+
+        #[cfg(feature = "pbr_multi_layer_material_textures")]
+        {
+            if self.clearcoat_texture.is_some() {
+                flags |= StandardMaterialFlags::CLEARCOAT_TEXTURE;
+            }
```
nitpick - these flags should probably be added in the prepass as well in case anybody wants to use them

pcwalton replied:
I don't know what this means. The prepass doesn't care about any of these flags, and we don't do this anywhere else (grepping for e.g. `BASE_COLOR` brings up nothing related to prepasses, except for alpha discard, which isn't relevant here).

JMS55 replied:
Yeah I wouldn't add them. Sometime soon we should revisit how pipeline keys are managed however.

robtfm reviewed (commented):
just a first pass. the changes to the core lighting functions are pretty extensive, @mockersf could we please run a demo comparison? 

i think we need to make proper efforts to stop the lighting code becoming more write-only than it already is. not saying this makes it much worse than it already was, but it's honestly very hard to follow.

it would also be nice to improve the example a bit (the three.js one is nicer, and i don't think their implementation is better) but that can certainly be done later.

pcwalton commented:
@robtfm How would you like me to improve the example? I wanted to include a transparent one because @JMS55 asked me how clearcoat looked with transparency. Should I add more of a noise pattern to the car paint? Redo the golf ball texture?

pcwalton commented:
@robtfm I tried to improve the lighting code as I went and make it as easy to follow as possible. Sorry to hear that I didn't succeed :( Do you have any suggestions?

robtfm commented:
> @robtfm How would you like me to improve the example? I wanted to include a transparent one because @JMS55 asked me how clearcoat looked with transparency. Should I add more of a noise pattern to the car paint? Redo the golf ball texture?

i think the main issue i have is the specular aliasing from the normal maps. it seems less "bitty" in the 3js version. perhaps just downres the normal map, or just use a smaller one maybe so the single pixel highlights are less frequent?
the texture on the carpaint is nicer in 3js as well (and the spinning spheres) but i'm not sure that's worth adding a new texture for.



> @robtfm I tried to improve the lighting code as I went and make it as easy to follow as possible. Sorry to hear that I didn't succeed :( Do you have any suggestions?

yes sorry, i didn't mean to sound so critical. i think if you hadn't made such an effort it would be much worse, whereas now it's not much worse, and with more functionality. the 2 specific comments about a param struct and factoring the meshlet sample were the main things i thought would help. i guess if the lighting functions api is clean then at least the trickier stuff is more contained.

pcwalton commented:
@robtfm I believe every comment has been addressed. `pbr_lighting.wgsl` has been largely rewritten. I made the spheres rotate, added some blur and decreased the levels on the scratch normal map, added some subtle blue noise as a normal map to the car paint, and switched to ACES filmic tonemapping, as this is what makes the normal map on the car paint in three.js appear purple.

![Screenshot 2024-04-22 160935](https://github.com/bevyengine/bevy/assets/157897/29d36b15-c446-4fca-9945-d40bdbc17995)

mockersf commented:
> just a first pass. the changes to the core lighting functions are pretty extensive, @mockersf could we please run a demo comparison?

Mostly good, just a change in the transmission example: 
![](https://pixel-eagle.vleue.com/project/B25A040A-A980-4602-B90C-D480AB84076D/run/852/compare/840?screenshot=3D%20Rendering/transmission.png)

pcwalton commented:
How do I get rid of the GitHub typos check on `Fo`?

mockersf commented:
add it to https://github.com/bevyengine/bevy/blob/main/typos.toml

pcwalton commented:
I *think* I fixed the transmission example, but it's hard for me to check locally as it's animated.

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.wgsl:
```diff
@@ -153,13 +161,35 @@ fn environment_map_light(
+    let Fc = F_Schlick_vec(clearcoat_F0, 1.0, clearcoat_NdotV) * clearcoat_strength;
+    let inv_Fc = 1.0 - Fc;
+
+    // We have to copy `clearcoat` here to work around a Naga bug.
```
What bug? Please open an issue with wgpu and link to it here.

pcwalton replied:
https://github.com/gfx-rs/wgpu/issues/5593, added a link

robtfm commented:
the gold and golf balls looks great now, but the blue sphere doesn't show: `ERROR bevy_asset::server: Path not found: [...]/assets/textures/BlueNoise-Normal.png`

pcwalton commented:
@robtfm Oops, forgot to check that file in! Fixed.

JMS55 commented on crates/bevy_gltf/Cargo.toml:
```diff
@@ -11,6 +11,7 @@ keywords = ["bevy"]
 [features]
 dds = []
 pbr_transmission_textures = []
+pbr_multi_layer_material_textures = []
```
I feel like we should split StandardMaterial into 2 / features / something. A basic one with the typical PBR stuff for lower end platforms, and one with all features included for newer platforms. Doesn't have to be this PR.

pcwalton replied:
Yeah, I considered doing that here, but decided against it because of the precedent set by transmission, which is also a more advanced feature. My personal feeling is that maybe `StandardMaterial` should be the glTF standard with no extensions, and then we can have `GltfExtensionMaterial`, `OpenPbrMaterial`, `MaterialXMaterial` that build on top of it. But I imagine there will be a lot of discussion around this topic, so that's for a followup :)

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.wgsl:
```diff
@@ -153,13 +162,36 @@ fn environment_map_light(
+
+    // Clearcoat
+
+#ifdef STANDARD_MATERIAL_CLEARCOAT
```
This file is getting rough 😅 . I won't block on it, but I'd prefer to start factoring it out in multiple functions with this file/other files.

pcwalton replied:
It's only 198 lines, which is pretty short. How were you thinking about splitting it up?

JMS55 replied:
Splitting the clear coat stuff up to a separate function would help maybe. You're right it's not so bad though (I was looking at it in the diff view).

pcwalton replied:
OK, done.

JMS55 commented on crates/bevy_pbr/src/pbr_material.rs:
```diff
@@ -338,6 +338,61 @@ pub struct StandardMaterial {
     pub occlusion_texture: Option<Handle<Image>>,
 
+    /// An extra thin translucent layer on top of the main PBR layer. According
+    /// to the Filament spec, this is commonly used for materials such as car
```
I would remove the mention of filament.

pcwalton replied:
Yeah, I just mentioned Filament there because I lifted the line from its documentation and was worried about plagiarism. I switched it to just "This is typically used for painted surfaces."

JMS55 commented on crates/bevy_pbr/src/pbr_material.rs:
```diff
@@ -338,6 +338,61 @@ pub struct StandardMaterial {
+    pub clearcoat_perceptual_roughness: f32,
+
+    /// An image texture that specifies the roughness of the clearcoat level in
+    /// the green channel.  Values from this texture are multiplied by the main
```
```suggestion
    /// the green channel. Values from this texture are multiplied by the main
```

JMS55 commented on crates/bevy_pbr/src/render/pbr_bindings.wgsl:
```diff
@@ -23,3 +23,11 @@
+@group(2) @binding(21) var clearcoat_roughness_texture: texture_2d<f32>;
+@group(2) @binding(22) var clearcoat_roughness_sampler: sampler;
+@group(2) @binding(23) var clearcoat_normal_texture: texture_2d<f32>;
+@group(2) @binding(24) var clearcoat_normal_sampler: sampler;
```
The continued per-texture sampler is rough, but I don't think there's a better solution until we have bindless and can use indices into a small array of samplers.

pcwalton replied:
Yeah, I really don't like it either. (In general glTF may have gone overboard with all the textures at the standards level; if you're really getting that exact about e.g. clearcoat roughness you probably want a node graph like MaterialX anyway. But standards are standards, so I added support anyway...)

JMS55 commented on crates/bevy_pbr/src/render/pbr_types.wgsl:
```diff
@@ -15,6 +15,8 @@ struct StandardMaterial {
     thickness: f32,
     ior: f32,
     attenuation_distance: f32,
+    clearcoat: f32,
```
You need to update standard_material_new() in this file too.

robtfm commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -79,6 +79,16 @@ fn pbr_input_from_standard_material(
+    let ddx_uv = in.ddx_uv;
+    let ddy_uv = in.ddy_uv;
+#else   // MESHLET_MESH_MATERIAL_PASS
+    let ddx_uv = vec2<f32>(0.0);
```
perhaps it would be better if `sample_texture` took a struct who's declaration varied based on the MESHLET define?

not a big deal though

pcwalton replied:
Done.

robtfm commented on crates/bevy_pbr/src/light_probe/environment_map.wgsl:
```diff
@@ -176,14 +177,13 @@ fn environment_map_light(
-        found_diffuse_indirect);
+
+    // We have to copy `clearcoat` here to work around Naga bug
+    // <https://github.com/gfx-rs/wgpu/issues/5593>.
```
from the duped issue this doesn't sound like a bug, just an unsupported extension (`unrestricted_pointer_parameters`).

not clear to me if this is an extension we can request / something that's supported in some contexts though.

robtfm replied:
Could we make this work by passing around an actual struct that contains pointers to a base, a base_layer and a clearcoat_layer?

pcwalton replied:
You can't put `ptr<function, ...>` types in structs in WGSL. I just tried.

robtfm replied:
Right, that is explicitly denied in the spec. Maybe use a fixed-size array of layers and pass the full ptr and an index into the layer-based functions..?

I don’t know if this would be worthwhile; I can try it and profile later if you don’t want to (I realise I’ve been pretty demanding already)

pcwalton replied:
OK, I did that (layers and indices).

robtfm commented on crates/bevy_pbr/src/render/pbr_lighting.wgsl:
```diff
@@ -40,8 +40,57 @@
-// Calculations for both diffuse and specular lights.
+// Input to a lighting function for a single layer (either the base layer or the
+// clearcoat layer).
+struct LayerLightingInput {
```
we might benefit from reordering fields here - i think the f32 ahead of the 16b aligned vec3s wastes some space. these are not i/o structs so i don't know how important it is.

pcwalton replied:
Reordered.

robtfm reviewed (commented):
i think it's a good improvement, though passing by reference really needs some ergonomics improvements (why no `->` wgsl?!).

JMS55 commented on crates/bevy_pbr/src/render/pbr_lighting.wgsl:
```diff
@@ -74,62 +136,140 @@ fn D_GGX(roughness: f32, NoH: f32, h: vec3<f32>) -> f32 {
+    light_position_radius: f32,
+) -> vec4<f32> {
+    // Unpack.
+    let R = (*input).R;
```
It's in the webgpu spec now, just not implemented by naga yet. "pointer composite access" https://gpuweb.github.io/gpuweb/wgsl/#composite-value-decomposition-expr

pcwalton commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -197,15 +205,58 @@ fn apply_pbr_lighting(
 
+    // Pack all the values into a structure.
+    var lighting_input: lighting::LightingInput;
+    lighting_input.base.NdotV = NdotV;
```
I would guess that any shader compiler worth its salt is going to inline and get rid of the struct anyway.

pcwalton replied:
I checked the RDNA disassembly and it is inlined and the structs don't exist at runtime. I haven't checked NVIDIA and Intel but I suspect it's the same. I'm going to go ahead and resolve this.

robtfm commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -402,17 +454,30 @@ fn apply_pbr_lighting(
-        false);
+    var transmissive_environment_light_input: lighting::LightingInput;
+    transmissive_environment_light_input.diffuse_color = vec3(1.0);
+    transmissive_environment_light_input.NdotV = 1.0;
```
i tried to get this to fail, but couldn't - finally realized it's because `STANDARD_MATERIAL_SPECULAR_OR_DIFFUSE_TRANSMISSION` is never defined.

pcwalton replied:
Fixed.

robtfm reviewed (approved):
looks good!

we should probably add something to the migration guide mentioning the changes required for custom prepasses/shaders that use `apply_normal_mapping`.

superdump commented:
I’ve read the discussion, has any GPU-profiling/benchmarking been done across vendors to see what impact the shader refactoring has on performance without clear coat?

pcwalton commented:
On NVIDIA, warp occupancy and register count are identical both before and after this PR.

![Screenshot 2024-05-01 150843](https://github.com/bevyengine/bevy/assets/157897/aa7f6737-3073-4d2c-9530-ab2ea789dc6e)

![Screenshot 2024-05-01 151432](https://github.com/bevyengine/bevy/assets/157897/2c294445-fa7d-4243-b987-214271d2c32c)

NthTensor commented:
No obvious performance differences for me testing vs main. What's blocking this moving forward? The more general changes seem like a marked improvement in standardization and code quality.

pcwalton commented:
@NthTensor It only has one approval. Having a second approval would be very helpful.

pcwalton commented:
Closing as I am burned out and have no motivation to get this landed anymore.

alice-i-cecile commented:
@pcwalton once merge conflicts are resolved this should be good to merge. Do reach out if you'd like a hand with that.

pcwalton commented:
Updated.

alice-i-cecile commented:
I'm comfortable with the thoroughness and expertise of the reviews here, and this isn't a controversial addition. Merging.
